kingpin/actors/packagecloud.py

```diff
@@ -0,0 +1,400 @@
+            """
+        pattern = re.compile(packages_to_delete)
+        packages_list_to_delete = [package['name'] for package in all_packages
+                                   if pattern.match(package['name'])]
+        packages_list_to_delete = set(packages_list_to_delete)
+        raise gen.Return(packages_list_to_delete)
+
+    @gen.coroutine
+    def _delete(self, packages_to_delete, repo, older_than=0,
+                number_to_keep=0):
+        """Generic packagecloud delete method, optionally supporting deleting
+        old packages by date and/or keeping a certain number of packages.
+
+        Args:
+            packages_to_delete: Regex of packages to delete, e.g. pkg1|pkg2
+            repo: name of the packagecloud repo to delete from
+            older_than: Delete packages created before this number of seconds
+            number_to_keep: Keep at least this number of each package
+
+        Returns:
+            A list of the packages that were deleted
+        """
+        all_packages = yield self._get_all_packages(repo=repo)
+        packages_list_to_delete = yield self._get_packages_list_to_delete(
+            packages_to_delete, all_packages)
+        deleted_packages = []
+
+        # Loop through each unique package to delete
+        for name in packages_list_to_delete:
+            package_versions = yield self._get_package_versions(
+                name, all_packages)
+
+            # Delete individual packages if they meet our criteria
+            number_in_repo = len(package_versions)
+            for package in package_versions:
+                if number_in_repo > number_to_keep:
+                    if older_than:
+                        if datetime.datetime.now() - package['created_at'] > \
+                           datetime.timedelta(seconds=older_than):
+                            pass
+                        else:
+                            continue
+
+                    msg = '%s/%s/%s' % (
+                        repo, package['distro_version'], package['filename'])
+                    if self._dry:
+                        log.info('Would have deleted %s' % msg)
+                    else:
+                        log.info('Deleting %s' % msg)
+
+                        yield self._packagecloud_client.delete(
+                            token=TOKEN, account=ACCOUNT, repo=repo,
+                            distro_version=package['distro_version'],
+                            filename=package['filename']
+                        ).http_delete()
+
+                    # Decrement list of packages to track how many are left
+                    number_in_repo = number_in_repo - 1
+                    # Keeping track of removed packages for return value
+                    deleted_packages.append(package)
+
+        raise gen.Return(deleted_packages)
+
+
+class Delete(PackagecloudBase):
+
+    """packagecloud Delete Actor Class
```

Comment:
Clean up the doc string here .. .the docs will actually be generated from this string. Consider something like:

```
Deletes old package versions from a PackageCloud repo.

... write up a full description of the behavior here. How the logic works with `number_to_keep` and `packages_to_delete`, etc.
```